File: javascript_code/graph/weighted/Dijkstra1.js

// 时间复杂度：O(V * (logV + E))
export class Dijkstra1 {

    constructor(g, source, minHeap) {
        this.g = g

        const MAX_INT = Math.pow(2, 31) - 1

        this.distance = new Array(g.getV()).fill(MAX_INT)
        this.distance[source] = 0
        this.visited = new Array(g.getV()).fill(false)

        minHeap.add([source, 0])
        while (!minHeap.isEmpty()) { // O(V)
            // 1. 找到当前没有访问的最短路径节点
            const curr = minHeap.remove()[0] // O(logV)
            if (this.visited[curr]) continue

            // 2. 确认这个节点的最短路径就是当前大小
            this.visited[curr] = true

            // 3. 根据这个节点的最短路径大小，更新其他节点的路径长度
            for (const w of g.adj(curr)) { // O(E)
                if (!this.visited[w]) {
                    if (this.distance[curr] + g.getWeight(curr, w) < this.distance[w]) {
                        this.distance[w] = this.distance[curr] + g.getWeight(curr, w)
                        minHeap.add([w, this.distance[w]])
                    }
                }
            }
        }
    }

    minDistanceTo(v) {
        this.validateVertex(v)
        return this.distance[v]
    }

    validateVertex(v) {
        if (v < 0 && v >= this.g.getV()) {
            throw new Error("顶点不合法，超出范围")
        }
    }
}

File: python_code/graph/weighted/Dijkstra1.py
import heapq

from graph.weighted.WeightAdjSet import WeightAdjSet


# 时间复杂度：O(V * (logV + E))
class Dijkstra1:
    def __init__(self, g: WeightAdjSet, source: int):
        self.g = g
        self.distance = [2**31 - 1] * g.getV()
        self.distance[source] = 0

        visited = [False] * g.getV()
        # 小顶堆，按照 dis 比较，所以 pq 中存储的是 (dis, v)
        pq = []
        heapq.heappush(pq, (0, source))
        while len(pq):
            # 1. 找到当前没有访问的最短路径节点
            curr = heapq.heappop(pq)[1]
            if visited[curr]:
                continue

            # 2. 确认这个节点的最短路径就是当前大小
            visited[curr] = True

            # 根据这个节点的最短路径大小，更新其他节点的路径长度
            for w in g.adj(curr):
                if not visited[w]:
                    if self.distance[curr] + g.getWeight(curr, w) < self.distance[w]:
                        self.distance[w] = self.distance[curr] + g.getWeight(curr, w)
                        heapq.heappush(pq, (self.distance[w], w))

    # 求 source 到 v 的最短路径
    def minDistanceTo(self, v):
        self.validateVertex(v)
        return self.distance[v]

    def validateVertex(self, v):
        if v < 0 or v >= self.g.getV():
            raise Exception("顶点 {} 不合格".format(v))


if __name__ == '__main__':
    g = WeightAdjSet("C:/workspace/douma-algo/data/Dijkstra.txt")
    dijkstra = Dijkstra1(g, 0)
    print(dijkstra.minDistanceTo(1))